### PR TITLE
Cleanup: Check packages in unnamed module when creating module

### PR DESCRIPTION
**Pull request by theresa-m**
Fixes: #6462
Original pr: https://github.com/eclipse/openj9/pull/6709

Signed-off-by: Theresa Mammarella <Theresa.T.Mammarella@ibm.com>

### CODE REVIEW

**Comment by DanHeidinga** [comment]
> pending startup performance testing.

@theresa-m Do you have performance results for this?

**Comment by theresa-m** [comment]
Yes. Daytrader startup is showing a 6% slowdown.

```
Benchmark mean Daytrader startup: 4612 ms (CI 0.0052)
With changes mean Daytrader startup: 4892 ms (CI 0.0435)

% difference = (4892/4612)*100 = 106%
```

I will work on getting some profiling results.

**Comment by theresa-m** [comment]
Instead of having two linear searches (one in package hash table and one in class hash table) I replaced the package table search with the class table search. This resulted in a negligible performance impact.

File: runtime/j9vm/java11vmi.c [hunk 1]
```diff
@@ -457,16 +458,27 @@ static BOOLEAN
...
 	if (NULL != packages) {
 		U_32 const arrayLength = numPackages;
 		if (0 != arrayLength) {
 			U_32 i = 0;
-			for (i = 0; i < arrayLength; i++) {
+			for (i = 0; success && (i < arrayLength); i++) {
 				const char *packageName = packages[i];
-				if (isPackageDefined(currentThread, classLoader, packageName)) {
+				if (checkDefinedPackages
+						&& vmFuncs->isAnyClassLoadedFromPackage(classLoader, (U_8*) packageName, strlen(packageName))
```
Comment (DanHeidinga) [inline runtime/j9vm/java11vmi.c]: Changing from `isPackageDefined` -> `isAnyClassLoadedFromPackage` - have you confirmed this gives the same results?  My read of the two functions indicates they'd give different results if the module hasn't loaded a class in the package yet.  Basically, two modules could "claim" `package.A` with this code provided they hadn't loaded a class in it yet, splitting the package across the modules.

Do you have a test case for this case?

**Comment by pshipton** [comment]
We're past Milestone 2 for the 0.20.0 release, moving this forward to the next release.

File: runtime/j9vm/java11vmi.c (hunk 1 above)
Comment (DanHeidinga) [reply runtime/j9vm/java11vmi.c]: @theresa-m I was circling back to this as I thought I saw similar code in a JDK 15 PR.  Can you confirm the test case for this?

Comment (theresa-m) [reply runtime/j9vm/java11vmi.c]: I tried your suggested scenario. OpenJ9 does throw the correct error, but as you predicted it's not in this line of code.

The case is covered later when class loaders are initialized:
```
java.lang.LayerInstantiationException: Package com.greetings in both module org.astro and module com.greetings
	at java.base/jdk.internal.module.ModuleBootstrap.checkSplitPackages(ModuleBootstrap.java:499)
	at java.base/jdk.internal.module.ModuleBootstrap.boot(ModuleBootstrap.java:406)
	at java.base/java.lang.ClassLoader.initializeClassLoaders(ClassLoader.java:211)
	at java.base/java.lang.Thread.initialize(Thread.java:430)
	at java.base/java.lang.Thread.<init>(Thread.java:155)
```

Comment (theresa-m) [reply runtime/j9vm/java11vmi.c]: Hey @DanHeidinga do you think the above error is sufficient to resolve this?

Comment (DanHeidinga) [reply runtime/j9vm/java11vmi.c]: Does it match the RI's behaviour?

Comment (theresa-m) [reply runtime/j9vm/java11vmi.c]: Yes it matches.

Comment (theresa-m) [reply runtime/j9vm/java11vmi.c]: @DanHeidinga I think this is ready to go unless there are other concerns.

**Comment by DanHeidinga** [comment]
I'm going to approve this and merge it into `master`.  I'd like to soak for a bit before we put it in a release and so have retargeted to v0.22.0

**Comment by DanHeidinga** [comment]
You've demonstrated that the class library will detect package conflicts between modules in `jdk.internal.module.ModuleBootstrap.checkSplitPackages(...)`, at least for modules in the built in classloaders - aka boot / system / app loaders and the test implies its trying to validate if there are package conflicts with the unnamed module (aka classpath).

Which lets us resolve this item.

I don't have a complete understanding of how packages should be validated in general and it would be good to get sorted out.

Basically, what Is it required to check existing packages against other layers?
Or to put it another way, if this is about splitting packages we'd need to validate class loads of the same package from the class path fail once a module in the same layer has already claimed the package.

There's also some questions here about cross layer class loading where the classpath splits a package in one unnamed module while another claims the package.

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk11,jdk14

**Comment by pshipton** [comment]
If this fixes #6462 then we should un-exclude the openjdk java/lang/ModuleLayer/BasicLayerTest.java test.

Trying a grinder on this test.
https://ci.eclipse.org/openj9/view/Test/job/Grinder/927/

**Comment by pshipton** [comment]
There is still a failure with openjdk java/lang/ModuleLayer/BasicLayerTest.java on jdk14. I'll reopen https://github.com/eclipse/openj9/issues/6462
```
test BasicLayerTest.testPackageAlreadyInUnnamedModule(): failure
org.testng.TestException: 
Method BasicLayerTest.testPackageAlreadyInUnnamedModule()[pri:0, instance:BasicLayerTest@f75f83fd] should have thrown an exception of type class java.lang.LayerInstantiationException
	at org.testng.internal.ExpectedExceptionsHolder.noException(ExpectedExceptionsHolder.java:89)
	at org.testng.internal.Invoker.handleInvocationResults(Invoker.java:1384)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:677)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:821)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1131)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:773)
	at org.testng.TestRunner.run(TestRunner.java:623)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1185)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1110)
	at org.testng.TestNG.run(TestNG.java:1018)
	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:94)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:298)
	at java.base/java.lang.Thread.run(Thread.java:851)
```
